### DISCUSSION TITLE
QEMU VMs based on Docker Images

### DISCUSSION THREAD

**Discussion by ghost** [GNS3/gns3-registry#875] [Show and tell]
While I like the simplicity and performance of Docker VMs, they have some drawbacks when used within GNS3:

* Persistence - Only the persistent directories will be saved. Additionally installed software will be lost after restart.
* Host kernel - The Docker VM can only use the features available in the host kernel.
* Security - The isolation between the Docker VM and the host is not as good as with QEMU VMs.

Looking for a way to make QEMU VMs lighter and easier to setup, I found this article:
[Execute Docker Containers as QEMU MicroVMs](https://mergeboard.com/blog/2-qemu-microvm-docker/)

That inspires me to create something similar. I created very small QEMU VMs based on Docker Images. For more information see <https://github.com/b-ehlers/qemu-minivm> (deleted). The project contains two ready-to-use appliances, just download them and import them into GNS3.

Here a short video using them. The main startup time is needed by the DHCP client, the system itself is ready in less than 2 seconds.

https://github.com/GNS3/gns3-registry/assets/49040325/1da0e8a6-693e-435c-895a-2e24e5f7f8d1

**Comment by ghost** [comment]


**Reply by grossmj** [reply]
Thanks, I am gonna check this out soon!

**Comment by grossmj** [comment]
I finally took some time to test. This is really great to have the option to run lightweight VMs, they boot really quickly. I think we should add your appliances to the gns3-registry repo.

![Screenshot from 2024-05-20 15-08-44](https://github.com/GNS3/gns3-registry/assets/4752614/a0af1b70-9903-4ffd-a011-48dbcba50482)

> Sad enough the use of QEMU MicroVMs is not possible in GNS3. GNS3 doesn't offer the virtio-blk-device and virtio-net-device devices. Instead of extending GNS3 I used the concept with normal QEMU VMs.

I may have missed something... what do you mean GNS3 doesn't offer the virtio-blk-device and virtio-net-device devices? How could GNS3 be extended to support these?

**Comment by ghost** [comment]
As Docker images are a good alternative, I abandon it.

**Reply by grossmj** [reply]
There is still the problem of persistence and security with Docker VMs. I think we should give the option to users and add the appliances to the registry.

**Reply by ghost** [reply]
Sure, giving the user more options is nice. However, I still believe that the benefits do not outweigh the effort required to support these new applications. So I won't create/support these appliances.

But if anyone is interested, I am happy to hand it over.

**Reply by grossmj** [reply]
Please, can you hand it over to me. I will keep it for personal use or for later in case we decide to add these appliances in the registry. Thanks 👍

**Reply by ghost** [reply]
The repository is (temporarily) available at <https://github.com/b-ehlers/qemu-minivm>. In the release area I have added some binaries. They should make it simpler to start with it. So you should probably copy these binaries as well.

I see no need to keep this repository private, you can use it any way you like. It's just I don't want to maintain it, because I think it is too much hassle. But, of course, you are free to have a different opinion.

If you have any questions, feel free to contact me.

**Reply by grossmj** [reply]
Thanks, I have cloned the repo and downloaded the binaries 👍

**Reply by ghost** [reply]
Forgot, that I wanted to document the origin of the binaries in the initramfs (`kernel/ramdisk-template`), so you are able to update them.

I just did that, so please sync your fork.
